File: backend/app/api/v1/routes/audit_logs.py
from typing import List
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi_injector import Injected

from app.auth.dependencies import auth, permissions
from app.schemas.audit_log import AuditLogRead, AuditLogSearchParams, AuditLogSearchResult
from app.services.audit_logs import AuditLogService
from datetime import datetime
from app.core.permissions.constants import Permissions as P

router = APIRouter()

@router.get("/search", response_model=List[AuditLogSearchResult], dependencies=[
    Depends(auth),
    Depends(permissions(P.AuditLog.READ))
])
async def search(
    date_from: datetime = Query(None),
    date_to: datetime = Query(None),
    action: str = Query(None),
    table_name: str = Query(None),
    entity_id: UUID = Query(None),
    user: UUID = Query(None),
    limit:  int | None = Query(None, ge=1, le=500),
    offset: int | None = Query(None, ge=0),
    service: AuditLogService = Injected(AuditLogService),
):
    """
    Search audit logs with optional filters:
    - start_date: Filter logs from this date
    - end_date: Filter logs until this date
    - action: Filter by action type (Insert, Update, Delete)
    - table_name: Filter by table name
    - entity_id: Filter by record UUID
    - modified_by: Filter by user UUID who made the change
    """
    search_params = AuditLogSearchParams(
        start_date=date_from,
        end_date=date_to,
        action=action,
        table_name=table_name,
        entity_id=entity_id,
        modified_by=user,
        limit=limit,
        offset=offset,
    )
    return await service.search_audit_logs(search_params)

@router.get("/{log_id}", response_model=AuditLogRead, dependencies=[
    Depends(auth),
    Depends(permissions(P.AuditLog.READ))
])
async def get_by_id(
    log_id: int,
    service: AuditLogService = Injected(AuditLogService),
):
    """
    Get a specific audit log entry by ID.
    """
    log = await service.get_audit_log_by_id(log_id)
    if log is None:
        raise HTTPException(status_code=404, detail=f"Audit log with ID {log_id} not found")
    return log

File: frontend/src/services/auditLogs.ts
import { apiRequest } from "@/config/api";
import { AuditLog } from "@/interfaces/audit-log.interface";
import { User } from "@/interfaces/user.interface";

export const fetchAuditLogs = async (
  date_from: string,
  date_to: string,
  action: string,
  user: string,
  limit: number,
  offset: number
): Promise<AuditLog[]> => {
  try {
    const queryParams = new URLSearchParams();

    if (date_from) queryParams.append("date_from", date_from);
    if (date_to) queryParams.append("date_to", date_to);
    if (action) queryParams.append("action", action);
    if (user) queryParams.append("user", user);

    queryParams.append("limit",  String(limit));
    queryParams.append("offset", String(offset));

    const data = await apiRequest<AuditLog[]>(
      "GET",
      `/audit-logs/search?${queryParams.toString()}`
    );

    return data || [];
  } catch (error) {
    throw error;
  }
};

export const fetchAuditLogDetails = async (id: string): Promise<AuditLog> => {
  try {
    const data = await apiRequest<AuditLog>("GET", `/audit-logs/${id}`);
    if (!data) {
      throw new Error(`No audit log found with ID: ${id}`);
    }
    if (!data.json_changes) {
      throw new Error("No json_changes found in the audit log.");
    }
    return data;
  } catch (error) {
    throw error;
  }
};

export const fetchUsers = async (): Promise<User[]> => {
  try {
    const data = await apiRequest<User[]>("GET", "user/");
    return data || [];
  } catch (error) {
    throw error;
  }
};
